### ISSUE TITLE
Norm 1 of tensor - bfloat16 - rounding errors?

### ISSUE THREAD

**Issue by VictorSanh** [pytorch/pytorch#92164]
### 🐛 Describe the bug

Depending on whether I compute the norm 1 of a tensor in bfloat16 or in fp32, I will get drastically different norms. My best guess is that is due to bfloat16's rounding errors.

```
>>> A = torch.ones(5120, dtype=torch.bfloat16)
>>> A
tensor([1., 1., 1.,  ..., 1., 1., 1.], dtype=torch.bfloat16)
>>> A.float().norm(1)
tensor(5120.)
>>> A.norm(1)
tensor(256., dtype=torch.bfloat16)
```

How can we fix that discrepancy that does not involve converting to fp32 and then converting back to bf16?

For what it's worth, 5'120 is representable in bf16:
```
>>> torch.tensor([5120], dtype=torch.bfloat16)
tensor([5120.], dtype=torch.bfloat16)
```

### Versions

```
Collecting environment information...
PyTorch version: 1.12.1+cu102
Is debug build: False
CUDA used to build PyTorch: 10.2
ROCM used to build PyTorch: N/A

OS: Ubuntu 18.04.6 LTS (x86_64)
GCC version: (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
Clang version: Could not collect
CMake version: Could not collect
Libc version: glibc-2.27

Python version: 3.8.13 (default, Mar 28 2022, 11:38:47)  [GCC 7.5.0] (64-bit runtime)
Python platform: Linux-5.4.0-1097-gcp-x86_64-with-glibc2.17
Is CUDA available: True
CUDA runtime version: Could not collect
CUDA_MODULE_LOADING set to: 
GPU models and configuration: 
GPU 0: Tesla V100-SXM2-16GB
GPU 1: Tesla V100-SXM2-16GB
GPU 2: Tesla V100-SXM2-16GB
GPU 3: Tesla V100-SXM2-16GB

Nvidia driver version: 495.46
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] mypy-extensions==0.4.3
[pip3] numpy==1.23.3
[pip3] torch==1.12.1
[pip3] torchvision==0.13.1
[conda] numpy                     1.23.3                   pypi_0    pypi
[conda] torch                     1.12.1                   pypi_0    pypi
[conda] torchvision               0.13.1                   pypi_0    pypi
```

cc @ezyang @gchanan @zou3519

**Comment by janeyx99** [comment]
Hello! As you guessed, this is to be expected-->when you add 256 + 1 in bfloat16, you would get 256 again 😛

Taking a step back--what is your intended use case here? If you are attempting to speed up training by converting between dtypes, the recommendation is to use [auto mixed precision](https://pytorch.org/docs/stable/amp.html) which will automatically convert between dtypes for ops that won't be too affected by the lowering of precision.

**Comment by VictorSanh** [comment]
Thanks for the answer!

I am trying to log norm-1 of trainable parameters during a training. More specifically, I would like to log the norm 1 of a LayerNorm module (thus the repetition of 1s), so typically logging the result `ln.weight.norm(p=1).item()`.

I don't think it would be handled by amp, would it?

**Comment by janeyx99** [comment]
Ah, but why use dtype bfloat16 if the 1s are important?

**Comment by VictorSanh** [comment]
When I get the module for norm loggin in bf16 mixed precision training, the modules happen to be in bf16.

Based on your comments, I know now that I should convert the weights to float32 before calling `.norm(p=1)`!

I guess my question is whether there are plans to mitigate the problems caused by iterative additions and rounding errors directly in the `norm` computation.

**Comment by janeyx99** [comment]
Ah interesting--it does sound like you should convert the weights to a higher precision fp before calling norm :)

To answer your question explicitly, there are no immediate plans to make norm aware of the dtype it is working with. However, it is possible through writing one single kernel for norm with higher precision intermediates. cc'ing @ngimel for thoughts here.

**Comment by ngimel** [comment]
THis is not to be expected and comes from bad norm implementation on CPU. Cuda produces correct result. 
cc @jgong5 @mingfeima

**Comment by XiaobingSuper** [comment]
Yes, this is to be expected, there has an issue with norm reduce which doesn't use accumulate type.

**Comment by janeyx99** [comment]
Removing triage review as it's already in progress via intel!

**Comment by stgzr** [comment]
Any update to this issue? It still happens in `pytorch==1.14.0a0+410ce96`

**Comment by ZailiWang** [comment]
This is fixed in latest nightly build (20230909).

**Comment by XiaobingSuper** [comment]
Yes, it was fixed by https://github.com/pytorch/pytorch/pull/95166.
